Kata/Program.cs

```
@@ -0,0 +1,40 @@
+﻿using Kata.Services;
+
+namespace Kata;
+
+public class Program
+{
+    private IParserService? _parserService;
+    private ISumService? _sumService;
+
+    public static void Main(string[] args)
+    {
+        var p = new Program();
+        IParserService _parserService = new ParserService();
+        ISumService _sumService = new SumService();
+
+        var reader = new StreamReader("TestFile.txt");
+        var file = reader.ReadToEnd();
+        var sum = p.Add(file);
+
+        Console.WriteLine(sum);
+    }
+
+    private int Add(string numbers)
+    {
+        if (string.IsNullOrEmpty(numbers))
+        {
+            return 0;
+        }
+
+        var (delimiters, parsedNumbers) = _parserService!.ParseDelimitersAndNumbers(numbers);
+        var (negativeNumbers, validNumbers) = _sumService!.GetNegativesAndValidNumbers(parsedNumbers);
+
+        if (negativeNumbers.Any())

```

Didn't feel good nesting the throw inside of a method.. Felt better having it in the caller

---

Not necessarily, throwing inside another method is just fine as the typical approach is to let the exception bubble up and be handled higher in the call stack. The problem here is that checking the negatives and throwing an exception should not be a direct responsibility of an "Add" method - it should be doing what its name suggests: adding.